src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpMetricSerializer.cs

```diff
@@ -0,0 +1,352 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Diagnostics;
+using OpenTelemetry.Metrics;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Serializer;
+
+internal class ProtobufOtlpMetricSerializer
+{
+    private const int ReserveSizeForLength = 4;
+    private const int TraceIdSize = 16;
+    private const int SpanIdSize = 8;
+
+    private static readonly Stack<List<Metric>> MetricListPool = [];
+    private static readonly Dictionary<string, List<Metric>> ScopeMetricsList = [];
+
+    internal static int WriteMetricsData(byte[] buffer, int writePosition, in Resources.Resource? resource, in Batch<Metric> batch)
+    {
+        foreach (var metric in batch)
+        {
+            var metricName = metric.MeterName;
+            if (!ScopeMetricsList.TryGetValue(metricName, out var metrics))
+            {
+                metrics = MetricListPool.Count > 0 ? MetricListPool.Pop() : new List<Metric>();
+                ScopeMetricsList[metricName] = metrics;
+            }
+
+            metrics.Add(metric);
+        }
+
+        writePosition = WriteResourceMetrics(buffer, writePosition, resource, ScopeMetricsList);
+        ReturnMetricListToPool();
+
+        return writePosition;
+    }
+
+    private static void ReturnMetricListToPool()
+    {
+        if (ScopeMetricsList.Count != 0)
+        {
+            foreach (var entry in ScopeMetricsList)
+            {
+                entry.Value.Clear();
+                MetricListPool.Push(entry.Value);
+            }
+
+            ScopeMetricsList.Clear();
+        }
+    }
+
+    private static int WriteResourceMetrics(byte[] buffer, int writePosition, in Resources.Resource? resource, in Dictionary<string, List<Metric>> scopeMetrics)
```

Comment:
nit: You don't really need to specify `in` for anything which is a class/reference type. Just the struct/value types (`Batch<T>`, `MetricPoint`, etc.).